docs/source/extensions/loops.rst

```diff
@@ -0,0 +1,382 @@
+            self._generator = None
+
+        def on_run_start(self, batch, optimizers, batch_idx):
+            super().on_run_start(batch, optimizers, batch_idx)
+            assert self.trainer.lightning_module.automatic_optimization
+
+            # We request the generator once and save it for later
+            # so we can call next() on it.
+            self._generator = self._get_generator(batch, batch_idx, opt_idx=0)
+
+        def _get_generator(self, batch, batch_idx, opt_idx):
+            step_kwargs = _build_training_step_kwargs(
+                self.trainer.lightning_module,
+                self.trainer.optimizers,
+                batch,
+                batch_idx,
+                opt_idx,
+                hiddens=None,
+            )
+
+            # Here we are basically calling lightning_module.training_step()
+            # and this returns a generator! The training_step is handled by the
+            # accelerator to enable distributed training.
+            generator = self.trainer.accelerator.training_step(step_kwargs)
+            return generator
+
+        def _make_step_fn(self, batch, batch_idx, opt_idx):
+            return partial(self._training_step, self._generator)
+
+        def _training_step(self, generator):
+            lightning_module = self.trainer.lightning_module
+
+            # Here, instead of calling lightning_module.training_step()
+            # we call next() on the generator!
+            training_step_output = next(generator)
+
+            self.trainer.accelerator.post_training_step()
+            training_step_output = self.trainer.call_hook(
+                "training_step_end",
+                training_step_output,
+            )
+            result = ClosureResult.from_training_step_output(
+                training_step_output,
+                self.trainer.accumulate_grad_batches,
+            )
+            return result
+
+
+Here we subclass the existing :class:`~pytorch_lightning.loops.optimization.optimizer_loop.OptimizerLoop` and modify the way it interacts with the model's :code:`training_step`.
+The new loop is called :code:`YieldLoop` and contains a reference to the generator returned by the :code:`training_step`.
+On every new run (over the optimizers) we call the :code:`training_step` method on the LightningModule which is supposed to return a generator as it contains the :code:`yield` statements.
+There must be as many :code:`yield` statements as there are optimizers.
+
+The alternative to this example *manual optimization* where the same can be achieved, but with the generator loop we can still get all benefits of manual optimization without having to call backward or zero grad ourselves.
+
+Given this new loop definition, here is how you connect it to the :code:`Trainer`:
+
+.. code-block:: python
+
+    model = LitModel()
+    trainer = Trainer()
+
+    yield_loop = YieldLoop()
+    trainer.fit_loop.epoch_loop.batch_loop.connect(optimizer_loop=yield_loop)
```

Comment:
This is different than the explanation at the top, where you just did trainer.fit_loop(loop).

Can you explain why this is required? Or have a more basic example?